homework-submission/week2/main.js

```diff
@@ -0,0 +1,171 @@
+ *      - Button[add]
+ *    + TaskList
+ *      + TaskItem
+ *        - Text[description]
+ *        - Text[deadline]
+ *        - Button[delete]
+ */
+
+const tasks = [
+  {
+    id: 1,
+    description: 'Get out of bed',
+    deadline: '2017-09-11',
+    done: true
+  },
+  {
+    id: 2,
+    description: 'Brush teeth',
+    deadline: '2017-09-10',
+    done: false
+  },
+  {
+    id: 3,
+    description: 'Eat breakfast',
+    deadline: '2017-09-09',
+    done: false
+  }
+];
+
+// Material Icon component
+const Icon = ({ children }) => <i className="material-icons">{children}</i>;
+
+// Task item component
+const TaskItem = ({ task, onRemove, onToggleMarked }) => (
+  <StyledListItem>
+    <div onClick={() => onToggleMarked(task.id)} className={task.done ? 'done' : ''}>
+      {task.done && <Icon>check</Icon>}
+      <StyledText bold>{task.description}</StyledText>
+      <StyledText small>{task.deadline}</StyledText>
+    </div>
+    <div>
+      <StyledButton inverse theme="secondary" onClick={() => onRemove(task.id)}>
+        <Icon>delete</Icon>
+      </StyledButton>
+    </div>
+  </StyledListItem>
+);
+
+// Task list component
+const TaskList = props => {
+  const sortedTasks = props.tasks.sort((a, b) => new Date(a.deadline) - new Date(b.deadline));
+  const totalTasks = props.tasks.length;
+  const totalCompletedTasks = sortedTasks.filter(task => task.done).length;
+  const unCompletedTasks = props.tasks.filter(task => !task.done);
+  const completedTasks = props.tasks.filter(task => task.done);
+
+  return (
+    <React.Fragment>
+      <StyledTitle>
+        Sorted Task List
+        <StyledText small>({totalCompletedTasks + '/' + totalTasks})</StyledText>
+      </StyledTitle>
+      <nav>
+        {completedTasks.map(task => (
+          <TaskItem key={task.id} task={task} {...props} />
+        ))}
+        {unCompletedTasks.map(task => (
+          <TaskItem key={task.id} task={task} {...props} />
+        ))}
+      </nav>
+    </React.Fragment>
+  );
+};
+
+// Task add form component
+const TaskAddForm = ({ onAdd }) => {
+  const handleSubmit = e => {
+    e.preventDefault();
+    const description = e.target.description.value;
+    const deadline = e.target.deadline.value;
+
+    onAdd(description, deadline);
+  };
+
+  // Source: https://css-tricks.com/prefilling-date-input/
+  const today = new Date().toISOString().substr(0, 10);
+
+  return (
+    <form name="form" onSubmit={handleSubmit}>
+      <StyledInput name="description" type="text" placeholder="Description" />
+      <StyledInput name="deadline" type="text" placeholder="Deadline" onFocus={e => (e.target.value = today)} />
+      <StyledButton full theme="primary">
+        <Icon>add</Icon>
+      </StyledButton>
+    </form>
+  );
+};
+
+// Task container component
+class TaskContainer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      tasks: this.props.data,
+      lastId: this.props.data.length
```

Comment:
since lastId can be calculated from tasks (as you already do), so this really doesn't need to be in the state.
whenever you need it just calculate the length of state.tasks

Reply:
Yes, it's true. Here I just wanted for all current and deleted previous todos to have a unique ID. Just a fake `AUTO_INCREMENT` function which exist in databases 🤷‍♂️